core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -642,9 +672,95 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
+                    if (inFlightBatch.inFlightState.state == RecordState.ACKNOWLEDGED ||
+                            inFlightBatch.inFlightState.state == RecordState.ARCHIVED) {
+                        lastOffsetBeforeStartOffset = inFlightBatch.lastOffset;
+                    } else {
+                        break;
+                    }
+                } else {
+                    boolean toBreak = false;
+                    for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                        if (offsetState.getValue().state == RecordState.ACKNOWLEDGED ||
+                                offsetState.getValue().state == RecordState.ARCHIVED) {
+                            lastOffsetBeforeStartOffset = offsetState.getKey();
+                        } else {
+                            toBreak = true;
+                            break;
+                        }
+                    }
+                    if (toBreak) break;
+                }
+            }
+
+            // If the acknowledgement failed, then we cannot move the startOffset
+            if (lastOffsetBeforeStartOffset != -1) {
+                // This is true if all records in the cachedState have been acknowledged (either Accept or Reject).
+                if (lastOffsetBeforeStartOffset == cachedState.lastEntry().getValue().lastOffset()) {
+                    startOffset = -1;
+                    endOffset = -1;
+                    cachedState.clear();
+                } else {
+                    long firstKeyToRemoveFromCachedState = cachedState.firstKey();
+                    long lastKeyToRemoveFromCachedState;
+                    NavigableMap.Entry<Long, InFlightBatch> entry = cachedState.floorEntry(lastOffsetBeforeStartOffset);
+                    if (lastOffsetBeforeStartOffset == entry.getValue().lastOffset) {
+                        startOffset = cachedState.higherKey(lastOffsetBeforeStartOffset);
+                        lastKeyToRemoveFromCachedState = entry.getKey();
+                    } else {
+                        startOffset = lastOffsetBeforeStartOffset + 1;
+                        if (entry.getKey().equals(cachedState.firstKey())) {
+                            lastKeyToRemoveFromCachedState = -1;
+                        } else {
+                            lastKeyToRemoveFromCachedState = cachedState.lowerKey(entry.getKey());
+                        }
+                    }
+
+                    // This is true when the first batch in the cachedState contains some unacknowledged records
+                    if (lastKeyToRemoveFromCachedState != -1) {
+                        NavigableMap<Long, InFlightBatch> subMap = cachedState.subMap(firstKeyToRemoveFromCachedState, true, lastKeyToRemoveFromCachedState, true);
+                        for (Long key : subMap.keySet()) {
+                            cachedState.remove(key);
+                        }
+                    }
```

Comment:
Are we cleaning up the per offset state from the first batch. Though fine but the cached batch baseOffset should also move accordingly else it will have negative impact on rest of the code. My initial suggestion to you was not to remove partial entries from batch. Either remove complete batch or not, as the max inflight limit is soft limit.